# Preliminary Steps

In [33]:
import pandas as pd
import numpy as np
import re
import os
import subprocess 

In [34]:
# Download and unzip shivamb/netflix-shows dataset
!kaggle datasets download shivamb/netflix-shows 

Dataset URL: https://www.kaggle.com/datasets/shivamb/netflix-shows
License(s): CC0-1.0
netflix-shows.zip: Skipping, found more recently modified local copy (use --force to force download)


In [35]:
# Only unzip if netflix-shows.zip hasn't been unzipped already
dir_files = os.listdir(os.getcwd())
unzipped_csvs = ['netflix-shows.csv']
if set(dir_files).isdisjoint(set(unzipped_csvs)) and "netflix-shows.zip" in dir_files:
    subprocess.run(["unzip", "-o", "netflix-shows.zip"], check=True)

Archive:  netflix-shows.zip


  inflating: netflix_titles.csv      


Try reading in the data on your own!

In [36]:
# TODO: Read your data as a pandas dataframe. Save the dataframe to a variable named "netflix"
DATA_PATH = os.path.join(os.getcwd(), "netflix_titles.csv")
netflix = pd.read_csv('netflix_titles.csv')
netflix

show_id     type                  title         director  \
0         s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show          Blood & Water              NaN   
2         s3  TV Show              Ganglands  Julien Leclercq   
3         s4  TV Show  Jailbirds New Orleans              NaN   
4         s5  TV Show           Kota Factory              NaN   
...      ...      ...                    ...              ...   
8802   s8803    Movie                 Zodiac    David Fincher   
8803   s8804  TV Show            Zombie Dumb              NaN   
8804   s8805    Movie             Zombieland  Ruben Fleischer   
8805   s8806    Movie                   Zoom     Peter Hewitt   
8806   s8807    Movie                 Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                   NaN            NaN   
4     Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   
...                                                 ...            ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

              date_added  release_year rating   duration  \
0     September 25, 2021          2020  PG-13     90 min   
1     September 24, 2021          2021  TV-MA  2 Seasons   
2     September 24, 2021          2021  TV-MA   1 Season   
3     September 24, 2021          2021  TV-MA   1 Season   
4     September 24, 2021          2021  TV-MA  2 Seasons   
...                  ...           ...    ...        ...   
8802   November 20, 2019          2007      R    158 min   
8803        July 1, 2019          2018  TV-Y7  2 Seasons   
8804    November 1, 2019          2009      R     88 min   
8805    January 11, 2020          2006     PG     88 min   
8806       March 2, 2019          2015  TV-14    111 min   

                                              listed_in  \
0                                         Documentaries   
1       International TV Shows, TV Dramas, TV Mysteries   
2     Crime TV Shows, International TV Shows, TV Act...   
3                                Docuseries, Reality TV   
4     International TV Shows, Romantic TV Shows, TV ...   
...                                                 ...   
8802                     Cult Movies, Dramas, Thrillers   
8803             Kids' TV, Korean TV Shows, TV Comedies   
8804                            Comedies, Horror Movies   
8805                 Children & Family Movies, Comedies   
8806     Dramas, International Movies, Music & Musicals   

                                            description  
0     As her father nears the end of his life, filmm...  
1     After crossing paths at a party, a Cape Town t...  
2     To protect his family from a powerful drug lor...  
3     Feuds, flirtations and toilet talk go down amo...  
4     In a city of coaching centers known to train I...  
...                                                 ...  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...  

[8807 rows x 12 columns]

# EDA

We have a much larger dataset than last week. What are all the columns and what do they mean? What does each row mean? Take a some time to look through the data and understand what we're working with.

Some useful functions: describe(), unique(), sort_values(), dtypes(), shape, columns, info(), isnull(), value_counts(), and more if you can think of them. Get used to exploring data like this so you know what you're working with first!

In [37]:
netflix.head(5)

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [38]:
netflix.tail(5)

show_id     type        title         director  \
8802   s8803    Movie       Zodiac    David Fincher   
8803   s8804  TV Show  Zombie Dumb              NaN   
8804   s8805    Movie   Zombieland  Ruben Fleischer   
8805   s8806    Movie         Zoom     Peter Hewitt   
8806   s8807    Movie       Zubaan      Mozez Singh   

                                                   cast        country  \
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   
8803                                                NaN            NaN   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

             date_added  release_year rating   duration  \
8802  November 20, 2019          2007      R    158 min   
8803       July 1, 2019          2018  TV-Y7  2 Seasons   
8804   November 1, 2019          2009      R     88 min   
8805   January 11, 2020          2006     PG     88 min   
8806      March 2, 2019          2015  TV-14    111 min   

                                           listed_in  \
8802                  Cult Movies, Dramas, Thrillers   
8803          Kids' TV, Korean TV Shows, TV Comedies   
8804                         Comedies, Horror Movies   
8805              Children & Family Movies, Comedies   
8806  Dramas, International Movies, Music & Musicals   

                                            description  
8802  A political cartoonist, a crime reporter and a...  
8803  While living alone in a spooky town, a young g...  
8804  Looking to survive in a world taken over by zo...  
8805  Dragged from civilian life, a former superhero...  
8806  A scrappy but poor boy worms his way into a ty...

In [39]:
netflix.shape
#df has 8807 rows, and 12 columns 

(8807, 12)

In [40]:
netflix.dtypes

show_id         object
type            object
title           object
director        object
cast            object
country         object
date_added      object
release_year     int64
rating          object
duration        object
listed_in       object
description     object
dtype: object

In [41]:
netflix.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [42]:
# TODO: Do your EDA here!
netflix.describe()

release_year
count   8807.000000
mean    2014.180198
std        8.819312
min     1925.000000
25%     2013.000000
50%     2017.000000
75%     2019.000000
max     2021.000000

In [43]:
## Checking for MISSING DATA

In [44]:
netflix.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [45]:
netflix.isnull().mean()

show_id         0.000000
type            0.000000
title           0.000000
director        0.299080
cast            0.093675
country         0.094357
date_added      0.001135
release_year    0.000000
rating          0.000454
duration        0.000341
listed_in       0.000000
description     0.000000
dtype: float64

In [46]:
netflix['duration'].value_counts()

duration
1 Season     1793
2 Seasons     425
3 Seasons     199
90 min        152
94 min        146
             ... 
16 min          1
186 min         1
193 min         1
189 min         1
191 min         1
Name: count, Length: 220, dtype: int64

In [47]:
netflix.columns

Index(['show_id', 'type', 'title', 'director', 'cast', 'country', 'date_added',
       'release_year', 'rating', 'duration', 'listed_in', 'description'],
      dtype='object')

In [48]:
netflix.duplicated().sum()

np.int64(0)

In [49]:
netflix['date_added'] = pd.to_datetime(netflix['date_added'], errors='coerce')
netflix.sort_values('date_added', ascending = True)

show_id     type                        title              director  \
5957   s5958    Movie         To and From New York  Sorin Dan Mihalcescu   
6611   s6612  TV Show              Dinner for Five                   NaN   
5956   s5957    Movie      Just Another Love Story          Ole Bornedal   
5955   s5956    Movie                     Splatter             Joe Dante   
7370   s7371    Movie  Mad Ron's Prevues from Hell            Jim Monaco   
...      ...      ...                          ...                   ...   
8539   s8540  TV Show                   The Tudors                   NaN   
8557   s8558  TV Show                The West Wing                   NaN   
8684   s8685  TV Show                      Vroomiz                   NaN   
8712   s8713  TV Show   Weird Wonders of the World                   NaN   
8755   s8756  TV Show            Women Behind Bars                   NaN   

                                                   cast  \
5957  Barbara King, Shaana Diya, John Krisiukenas, Y...   
6611                                                NaN   
5956  Anders W. Berthelsen, Rebecka Hemse, Nikolaj L...   
5955  Corey Feldman, Tony Todd, Tara Leigh, Erin Way...   
7370  Nick Pawlow, Jordu Schell, Jay Kushwara, Micha...   
...                                                 ...   
8539  Jonathan Rhys Meyers, Henry Cavill, James Frai...   
8557  Martin Sheen, Rob Lowe, Allison Janney, John S...   
8684  Joon-seok Song, Jeong-hwa Yang, Sang-hyun Um, ...   
8712                                      Chris Packham   
8755                                                NaN   

                                             country date_added  release_year  \
5957                                   United States 2008-01-01          2006   
6611                                   United States 2008-02-04          2007   
5956                                         Denmark 2009-05-05          2007   
5955                                   United States 2009-11-18          2009   
7370                                   United States 2010-11-01          1987   
...                                              ...        ...           ...   
8539  Ireland, Canada, United States, United Kingdom        NaT          2010   
8557                                   United States        NaT          2005   
8684                                     South Korea        NaT          2016   
8712                                  United Kingdom        NaT          2016   
8755                                   United States        NaT          2010   

     rating   duration                                          listed_in  \
5957  TV-MA     81 min              Dramas, Independent Movies, Thrillers   
6611  TV-MA   1 Season                       Stand-Up Comedy & Talk Shows   
5956  TV-MA    104 min                       Dramas, International Movies   
5955  TV-MA     29 min                                      Horror Movies   
7370     NR     84 min                         Cult Movies, Horror Movies   
...     ...        ...                                                ...   
8539  TV-MA  4 Seasons                                          TV Dramas   
8557  TV-14  7 Seasons                                          TV Dramas   
8684   TV-Y  3 Seasons                          Kids' TV, Korean TV Shows   
8712  TV-PG  2 Seasons  British TV Shows, Docuseries, Science & Nature TV   
8755  TV-14  3 Seasons                         Crime TV Shows, Docuseries   

                                            description  
5957  While covering a story in New York City, a Sea...  
6611  In each episode, four celebrities join host Jo...  
5956  When he causes a car accident that leaves a yo...  
5955  After committing suicide, a washed-up rocker r...  
7370  This collection cherry-picks trailers, forgott...  
...                                                 ...  
8539  All the splendor and scandal of England's 16th...  
8557  This powerful p

# Granularity, Scope, and Temporality

Once you have completed your EDA, you should be able to answer each of these broad ideas about the data. If not, you can always do more EDA!

*Note, you can jot down some quick notes instead of a long answer to help remind yourself about certain characteristics of this dataframe

#### Granularity: What information does each row give us? How is each row unique (i.e., what identifies each row? This is called the **"primary key"**)

**Insert answer here**
Each row gives us information about a different movie. For example, the first row gives us Dick Johnson is Dead, and provides information about that unique movie, and its filmaking properties. Each row is unique in the fact that it is a different movie for each row, giving data on various directors, ratings, type, and more.

#### Scope: What are some interesting things we could learn from the data (we can find this out by exploring the columns we have)? Do we have to manipulate the data in some way to get what we're interested in?

**Insert answer here**
frequency of movies posted during a certain time frame, average duration of movies within a time frame (are older movies longer/shorter on average), explore various genres, see whats the most popular, see the distrubution of release years, etc

#### Temporality: When was the data collected? How often is the data collected (if there is a pattern)? Do we need to adjust for consistency in the dates/times? Do we need to adjust the data type of our time variables?

**Insert answer here**
The data was firest collected in 2008, and the most recent movie title was shown in 2021. Date added does need to be adjusted. I used pd.set_datetime to format the dates properly, and also eventually, I would clean the dataset as well. 

We will be having people share their responses to each of these questions and discuss what everybody found in the data!

# Faithfulness

This is all about deciding whether you can trust the data in the form it came in, or whether you need to make adjustments to do so.

Did you find any strange or inconsistent values? Can you figure out how the data was collected? Are there any duplicate values (in this case, there shouldn't be because each movie is a separate one)?

**Insert answer here**
There are missing values, and for the directors column almost 30% of data is missing for that cateory. Furthermore, there were also values missing in other columsn but like a smaller chunk of the data I'm guessing this data was scraped from internet sources from Netflix maybe where movie descriptions, and basic details can easily be located. I checked for duplicate values, and there are none. 

Possibly one of the most important questions is what to do with default values. These can be NaN, NA, 0, or something else entirely. Based on what we see from this data, what should we do with our missing values, and should we take different steps for different columns? You can check the slides for some examples of what people do with missing data.

**Insert answer here**
I would change all the missing director values from null to 'unknown.' Director is useful but not essential for most like in general statistical explorations. For the date, there are only ten rows missing so I would drop them. Rating has 4 rows missing so I would drop, same with duration, and date-added. cast, and country had 10% missing which is a lot so I can either replace with 'Unknown' or drop the rows based on my needs. 

We'll discuss again afterwards!

# Variable Typing

Just for practice, it's important to be able to identify what type of variable something is just by looking at it. As a refresher, the 4 types are Quantitative Continous, Quantitative Discrete, Qualitative Ordinal, and Qualitative Nominal.

#### Question: What are the variables types of
a) Director

b) Country

c) Date Added

d) Release Year


**Insert Answer Here**

a Object
b Object
c Object
d Int

# Main Exercise

A common, but often misguided, way to handle datasets that have outliers/untrustworthy data is to drop all rows or columns that have missing values.


Find the shape of our dataset before we do any data manipulation and compare it with the shape after you use the dropna function. What do you notice and how many columns or rows were removed from our data?

Tip: You should make a deep copy of a dataframe by using {data name}.copy(deep=True). Give it an appropriate name to reflect the fact that this will be the dataframe that will have all null values dropped without any previous operations performed on it.

In [50]:
# TODO: Find the shape of our original data vs the shape of our data after we dropna
netflix.shape

(8807, 12)

In [51]:
netflix_no_nulls = netflix.dropna()
netflix_no_nulls.shape

(5328, 12)

You should've found that some of our data was removed after using the drop_na function. Let's go back to our original dataset before we used the function.

Try creating a list of the columns that have null values. Feel free to search things up on StackOverflow, the Pandas documentations, etc.

In [52]:
# TODO: Create a list of columns that have null values
cols_with_nulls = netflix.columns[netflix.isnull().any()].tolist()
cols_with_nulls


['director', 'cast', 'country', 'date_added', 'rating', 'duration']

Now that we have the columns where missing values are present, we have a better idea of what the data types of our missing values actually area.

Try filling in the missing director values with an empty string "". You can check if you did this operation correctly by checking the columns that have missing values and replacing the original dataframe name in the code with the name of the copy you should make for this task.

In [53]:
# TODO: Fill in missing director values with an emoty string
director_fill = netflix.copy()
director_fill['director'] = director_fill['director'].fillna("")


Let's check if filling the "director" column will help reduce the number of rows that will be removed when we use the dropna function.

Use the dropna function on the dataframe that is created after filling in the "director" column and find its shape. Compare the shape of this dataframe to the shape of the original dataframe and the dataframe after dropping all null values.

In [54]:
# TODO: Compare size of original data, data when dropping na after replacing missing directors with empty string, and data with just dropna
director_fill.shape

(8807, 12)

You should notice that the size of the dataframe that is created when we drop null values after filling in the "director" column will have less rows than the original dataframe, but more than the dataframe that is created when we just drop all values without performing any operations.

This should serve as a basic exercise on how you can approach data that is not 100% clean. Most data in consulting projects and in the real world will force you to find ways to balance the preservation of the original data and getting rid of the unecessary parts of the data.

### String Manipulation. We'll load in a new dataset for this part.

In [58]:
!kaggle datasets download -d kazanova/sentiment140

Dataset URL: https://www.kaggle.com/datasets/kazanova/sentiment140
License(s): other
100%|██████████████████████████████████████| 80.9M/80.9M [00:03<00:00, 24.2MB/s]



In [68]:
# TODO: load in the kazanova/sentiment140 data and unzip the file that is downloaded.
dir_files = os.listdir(os.getcwd())

# Expected file(s) after unzipping sentiment140.zip
unzipped_csvs = ["training.1600000.processed.noemoticon.csv"]

# If the CSV is not present but the zip is, unzip it
if set(dir_files).isdisjoint(set(unzipped_csvs)) and "sentiment140.zip" in dir_files:
    subprocess.run(["unzip", "-o", "sentiment140.zip"], check=True)

In [70]:
# TODO: Read the data from the csv file that is generated from unzipping as a dataframe.
# You should use the latin-1 encoding for this step. Name the dataframe tweets_data

tweets_data = pd.read_csv(
    "training.1600000.processed.noemoticon.csv",
    encoding="latin-1",
)
tweets_data.head()

0  1467810369  Mon Apr 06 22:19:45 PDT 2009  NO_QUERY _TheSpecialOne_  \
0  0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY   scotthamilton   
1  0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY        mattycus   
2  0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY         ElleCTF   
3  0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY          Karoli   
4  0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY        joy_wolf   

  @switchfoot http://twitpic.com/2y1zl - Awww, that's a bummer.  You shoulda got David Carr of Third Day to do it. ;D  
0  is upset that he can't update his Facebook by ...                                                                   
1  @Kenichan I dived many times for the ball. Man...                                                                   
2    my whole body feels itchy and like its on fire                                                                    
3  @nationwideclass no, it's not behaving at all....                                                                   
4                      @Kwesidei not the whole crew

In [71]:
# Some cleanup of the data, this cell should run correctly if you named and read the dataframe properly
tweets = tweets_data.drop(columns = ["NO_QUERY"]).rename(columns={"0": "Polarity",
                                                                  "1467810369": "id",
                                                                  "Mon Apr 06 22:19:45 PDT 2009" : "Date",
                                                                  "_TheSpecialOne_" : "Username"})
final_tweets = tweets.set_axis([*tweets.columns[:-1], 'Tweet'], axis=1)
final_tweets.head(5)

Polarity          id                          Date       Username  \
0         0  1467810672  Mon Apr 06 22:19:49 PDT 2009  scotthamilton   
1         0  1467810917  Mon Apr 06 22:19:53 PDT 2009       mattycus   
2         0  1467811184  Mon Apr 06 22:19:57 PDT 2009        ElleCTF   
3         0  1467811193  Mon Apr 06 22:19:57 PDT 2009         Karoli   
4         0  1467811372  Mon Apr 06 22:20:00 PDT 2009       joy_wolf   

                                               Tweet  
0  is upset that he can't update his Facebook by ...  
1  @Kenichan I dived many times for the ball. Man...  
2    my whole body feels itchy and like its on fire   
3  @nationwideclass no, it's not behaving at all....  
4                      @Kwesidei not the whole crew

#### Now that we have our **final_tweets** dataset, we can start working with it. Let's practice string methods first.

In [73]:
# TODO: Make all of the strings in the "Username" columns lowercase

final_tweets['Username'] = final_tweets['Username'].str.lower()
final_tweets

Polarity          id                          Date         Username  \
0               0  1467810672  Mon Apr 06 22:19:49 PDT 2009    scotthamilton   
1               0  1467810917  Mon Apr 06 22:19:53 PDT 2009         mattycus   
2               0  1467811184  Mon Apr 06 22:19:57 PDT 2009          ellectf   
3               0  1467811193  Mon Apr 06 22:19:57 PDT 2009           karoli   
4               0  1467811372  Mon Apr 06 22:20:00 PDT 2009         joy_wolf   
...           ...         ...                           ...              ...   
1599994         4  2193601966  Tue Jun 16 08:40:49 PDT 2009  amandamarie1028   
1599995         4  2193601969  Tue Jun 16 08:40:49 PDT 2009      thewdboards   
1599996         4  2193601991  Tue Jun 16 08:40:49 PDT 2009           bpbabe   
1599997         4  2193602064  Tue Jun 16 08:40:49 PDT 2009     tinydiamondz   
1599998         4  2193602129  Tue Jun 16 08:40:50 PDT 2009   ryantrevmorris   

                                                     Tweet  
0        is upset that he can't update his Facebook by ...  
1        @Kenichan I dived many times for the ball. Man...  
2          my whole body feels itchy and like its on fire   
3        @nationwideclass no, it's not behaving at all....  
4                            @Kwesidei not the whole crew   
...                                                    ...  
1599994  Just woke up. Having no school is the best fee...  
1599995  TheWDB.com - Very cool to hear old Walt interv...  
1599996  Are you ready for your MoJo Makeover? Ask me f...  
1599997  Happy 38th Birthday to my boo of alll time!!! ...  
1599998  happy #charitytuesday @theNSPCC @SparksCharity...  

[1599999 rows x 5 columns]

#### Try making all the entries in the Tweet column uppercase and save this in a final_tweets_upper dataframe.

In [76]:
# TODO: make a copy of final_tweets and make the "tweets" column all uppercase

upper_tweets = final_tweets.copy()
upper_tweets['Tweet'] = upper_tweets['Tweet'].str.upper()
upper_tweets

Polarity          id                          Date         Username  \
0               0  1467810672  Mon Apr 06 22:19:49 PDT 2009    scotthamilton   
1               0  1467810917  Mon Apr 06 22:19:53 PDT 2009         mattycus   
2               0  1467811184  Mon Apr 06 22:19:57 PDT 2009          ellectf   
3               0  1467811193  Mon Apr 06 22:19:57 PDT 2009           karoli   
4               0  1467811372  Mon Apr 06 22:20:00 PDT 2009         joy_wolf   
...           ...         ...                           ...              ...   
1599994         4  2193601966  Tue Jun 16 08:40:49 PDT 2009  amandamarie1028   
1599995         4  2193601969  Tue Jun 16 08:40:49 PDT 2009      thewdboards   
1599996         4  2193601991  Tue Jun 16 08:40:49 PDT 2009           bpbabe   
1599997         4  2193602064  Tue Jun 16 08:40:49 PDT 2009     tinydiamondz   
1599998         4  2193602129  Tue Jun 16 08:40:50 PDT 2009   ryantrevmorris   

                                                     Tweet  
0        IS UPSET THAT HE CAN'T UPDATE HIS FACEBOOK BY ...  
1        @KENICHAN I DIVED MANY TIMES FOR THE BALL. MAN...  
2          MY WHOLE BODY FEELS ITCHY AND LIKE ITS ON FIRE   
3        @NATIONWIDECLASS NO, IT'S NOT BEHAVING AT ALL....  
4                            @KWESIDEI NOT THE WHOLE CREW   
...                                                    ...  
1599994  JUST WOKE UP. HAVING NO SCHOOL IS THE BEST FEE...  
1599995  THEWDB.COM - VERY COOL TO HEAR OLD WALT INTERV...  
1599996  ARE YOU READY FOR YOUR MOJO MAKEOVER? ASK ME F...  
1599997  HAPPY 38TH BIRTHDAY TO MY BOO OF ALLL TIME!!! ...  
1599998  HAPPY #CHARITYTUESDAY @THENSPCC @SPARKSCHARITY...  

[1599999 rows x 5 columns]

#### Now let's try to replace something in a string!

In [84]:
# TODO: create a dataframe that filters out the final_tweets dataframe's tweets and only shows tweets with underscores
tweets_with_underscores = final_tweets.copy()
tweets_with_underscores = final_tweets[final_tweets["Tweet"].str.contains("_", na=False)]
tweets_with_underscores

Polarity          id                          Date         Username  \
7               0  1467811795  Mon Apr 06 22:20:05 PDT 2009  2hood4hollywood   
21              0  1467814119  Mon Apr 06 22:20:40 PDT 2009        cooliodoc   
68              0  1467825084  Mon Apr 06 22:23:30 PDT 2009    presidentsnow   
108             0  1467838188  Mon Apr 06 22:26:54 PDT 2009      jess_higley   
118             0  1467839586  Mon Apr 06 22:27:18 PDT 2009        sonyolmos   
...           ...         ...                           ...              ...   
1599986         4  2193579092  Tue Jun 16 08:38:58 PDT 2009        cathriiin   
1599987         4  2193579191  Tue Jun 16 08:38:59 PDT 2009          tellman   
1599988         4  2193579211  Tue Jun 16 08:38:59 PDT 2009        jazzstixx   
1599992         4  2193579477  Tue Jun 16 08:39:00 PDT 2009      chloeamisha   
1599993         4  2193579489  Tue Jun 16 08:39:00 PDT 2009        evolvetom   

                                                     Tweet  
7                     @TATIANA_K NOPE THEY DIDN'T HAVE IT   
21        @ANGRY_BARISTA I BAKED YOU A CAKE BUT I ATED IT   
68       @LT_ALGONQUIN AGREED, I SAW THE FAILWHALE ALLL...  
108      @MARYKATHERINE_Q I KNOW! I HEARD IT THIS AFTER...  
118      @ERRE_SC AAW I MISS YA ALL TOO.. IM LEAVING TO...  
...                                                    ...  
1599986  @LA_R_A NEVEER  I THINK THAT YOU BOTH WILL GET...  
1599987  @ROY_EVERITT HA- GOOD JOB. THAT'S RIGHT - WE G...  
1599988                 @MS_HIP_HOP IM GLAD UR DOING WELL   
1599992                                  @SCOOBY_GRITBOYS   
1599993  @CLIFF_FORSTER YEAH, THAT DOES WORK BETTER THA...  

[75329 rows x 5 columns]

Task 2: In the cell below, use .str.replace to replace all the underscores in Username with a space (" ") in the tweeters_with__ dataframe. Output the changed tweeters_with__ dataframe.

Doc: https://pandas.pydata.org/docs/reference/api/pandas.Series.str.replace.html

In [ ]:
# TODO: replace all "_" with " " in the "Username" column
# tbl[col] when you want a column. Use tbl[tbl[col]] when you want to filter rows based on a condition in that column
user_with_under = final_tweets.copy()
user_with_under['Username'] = user_with_under['Username'].str.replace('_', " ")
user_with_under


# Regex

---
### Part 1 — Does the title start with a number? 

Some Netflix titles start with a number, like `"13 Reasons Why"` or `"21 Jump Street"`.

**Task:** Use `str.contains()` with a regex pattern to find all titles that **start with a digit**. Store the result in a new DataFrame called `number_titles` and print how many there are.

**Hint:** Should be 130


In [90]:
# TODO: write a pattern that matches titles starting with a digit
pattern = r"^\d"
number_titles = netflix[netflix["title"].str.contains(pattern, na=False)]
print(f"Titles starting with a digit: {len(number_titles)}")
number_titles["title"].head(10)

Titles starting with a digit: 130


188                 2 Alone in Paris
323                          30 Rock
324                          44 Cats
404    9to5: The Story of a Movement
438                 2 Weeks in Lagos
558                        6 Bullets
774                         2 Hearts
850                         99 Songs
851                 99 Songs (Tamil)
852                99 Songs (Telugu)
Name: title, dtype: object

---
### Part 2 — Search descriptions for key words using `\w` and `[ ]`

**Task:** Find all titles whose description mentions the word `"love"` **or** `"romance"` (case-insensitive). Count them and print 5 examples.

Then, as a follow-up, find titles whose description contains **a number followed immediately by the word `"day"` or `"days"`** — e.g. `"30 days"` or `"1 day"`.

**Hints:**
- Pass `case=False` to `str.contains()` to ignore capitalisation.
- For the number + day pattern: `\d+` matches the number, `\s` matches the space, `days?` matches `"day"` or `"days"` (the `?` makes the `s` optional).
- Answer should be 826 and 9 respectively


In [92]:
# TODO: descriptions mentioning love or romance (case-insensitive)
pattern_love = r"\b(love|romance)\b"  # <-- your pattern here

love_titles = netflix[netflix["description"].str.contains(pattern_love, case=False, na=False)]
print(f"Titles mentioning love or romance: {len(love_titles)}")
love_titles["title"].head(5)

Titles mentioning love or romance: 613


/var/folders/sd/w28b23ln2tb0w0c1qlj_hy0r0000gn/T/ipykernel_8121/1247452149.py:4: UserWarning: This pattern is interpreted as a regular expression, and has match groups. To actually get the groups, use str.extract.
  love_titles = netflix[netflix["description"].str.contains(pattern_love, case=False, na=False)]


25     Love on the Spectrum
26           Minsara Kanavu
30          Ankahi Kahaniya
47    The Smart Money Woman
93                     JJ+E
Name: title, dtype: object

In [93]:
# TODO: descriptions mentioning a number followed by "day" or "days"
pattern_days = r"^d+\s+day|days\b" 

days_titles = netflix[netflix["description"].str.contains(pattern_days, case=False, na=False)]
print(f"Titles with 'N day(s)' in description: {len(days_titles)}")
days_titles[["title", "description"]].head(5)

Titles with 'N day(s)' in description: 69


title  \
23   Go! Go! Cory Carson: Chrissy Takes the Wheel   
522                                         Larva   
531                             Rise: Ini Kalilah   
654                                       #Selfie   
655                                    #Selfie 69   

                                           description  
23   From arcade games to sled days and hiccup cure...  
522  Two curious worms spend their days investigati...  
531  Inspired by Malaysia's 14th General Election, ...  
654  Two days before their final exams, three teen ...  
655  After a painful breakup, a trio of party-lovin...